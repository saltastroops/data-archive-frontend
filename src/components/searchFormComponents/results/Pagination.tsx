import * as React from "react";

/**
 * Whether the user is moving to the next or the previous page.
 */
export type PaginationDirection = "NEXT" | "PREVIOUS";

interface IPaginationProps {
  fetchPage: (
    startIndex: number,
    limit: number,
    direction: PaginationDirection
  ) => void;
  itemsOnCurrentPage: number;
  itemsPerPage: number;
  itemsTotal: number;
  startIndex: number;
}

/**
 * A component for offset-based pagination.
 *
 * The following properties need to be passed to the component.
 *
 * fetchPage:
 *     A method for fetching the next page. The method must expect a start index
 *     and a limit as its arguments.
 * itemsOnCurrentPage:
 *     The number of items on the current page. Unless the current page is the
 *     last page, this will be the same as the items per page.
 * itemsPerPage:
 *     The number of items to include on a page. For the last page the actual
 *     number of items may be less.
 * itemsTotal:
 *     The total number of items for all pages combined.
 * startIndex:
 *     The index of the first item on the current page. An index offset of 0,
 *     i.e. the start index for the first page is 0.
 */
const Pagination = (props: IPaginationProps) => {
  const {
    fetchPage,
    itemsOnCurrentPage,
    itemsPerPage,
    itemsTotal,
    startIndex
  } = props;
  const previousIndex = startIndex - itemsPerPage;
  const nextIndex = startIndex + itemsOnCurrentPage;

  return (
    <div>
      <button
        disabled={startIndex <= 0}
        className="pagination-previous"
        onClick={() => fetchPage(previousIndex, itemsPerPage, "PREVIOUS")}
      >
        Previous page
      </button>
      <a>
        <span className="pagination-ellipsis">{startIndex + 1}</span>
        <span className="pagination-ellipsis">&hellip;</span>
        <span className="pagination-ellipsis">
          {startIndex + itemsOnCurrentPage}
        </span>
        <span className="pagination-ellipsis">of</span>
        <span className="pagination-ellipsis">{itemsTotal}</span>
      </a>
      <button
        disabled={startIndex + itemsPerPage >= itemsTotal}
        className="pagination-next"
        onClick={() => fetchPage(nextIndex, itemsPerPage, "NEXT")}
      >
        Next page
      </button>
    </div>
  );
};

export default Pagination;
