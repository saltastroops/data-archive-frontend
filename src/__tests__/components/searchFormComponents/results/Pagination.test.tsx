import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import Pagination from "../../../../components/searchFormComponents/results/Pagination";

const fetchPage = jest.fn();

describe("Pagination", () => {
  afterEach(() => {
    fetchPage.mockReset();
  });

  it("should render correctly with correct props", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={100}
          itemsTotal={201}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    const wrapper2 = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={10}
          itemsPerPage={10}
          itemsTotal={201}
          startIndex={10}
        />
      </MockedProvider>
    );
    expect(toJson(wrapper2)).toMatchSnapshot();

    const wrapper3 = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={1000}
          itemsPerPage={1000}
          itemsTotal={201}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(toJson(wrapper3)).toMatchSnapshot();
  });

  it("should enable previous button when there is previous data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={10}
          itemsPerPage={10}
          itemsTotal={201}
          startIndex={11}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-previous"]')
        .first()
        .prop("disabled")
    ).toBeFalsy();
  });

  it("should disable previous button when no previous data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={100}
          itemsTotal={201}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-previous"]')
        .first()
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should enable next button when there is next data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={100}
          itemsTotal={201}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-next"]')
        .first()
        .prop("disabled")
    ).toBeFalsy();
  });

  it("should disable next button when there is no next data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={100}
          itemsTotal={100}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-next"]')
        .first()
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should disable both next and previous button when there is no next and previous data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={100}
          itemsTotal={100}
          startIndex={0}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-previous"]')
        .first()
        .prop("disabled")
    ).toBeTruthy();
    expect(
      wrapper
        .find('button[data-test="pagination-next"]')
        .first()
        .prop("disabled")
    ).toBeTruthy();
  });

  it("should not disable next and fetchPage button when there is next and previous data", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={10}
          itemsTotal={100}
          startIndex={11}
        />
      </MockedProvider>
    );
    expect(
      wrapper
        .find('button[data-test="pagination-previous"]')
        .first()
        .prop("disabled")
    ).toBeFalsy();
    expect(
      wrapper
        .find('button[data-test="pagination-next"]')
        .first()
        .prop("disabled")
    ).toBeFalsy();
  });

  it("should call fetchPage with 'PREVIOUS' when previous button is clicked", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={100}
          itemsPerPage={10}
          itemsTotal={100}
          startIndex={10}
        />
      </MockedProvider>
    );

    const previousButton = wrapper
      .find('button[data-test="pagination-previous"]')
      .first();
    previousButton.simulate("click");
    expect(fetchPage).toHaveBeenCalledWith(0, 10, "PREVIOUS");
  });

  it("should call fetchPage with 'NEXT' when next button is clicked", () => {
    const wrapper = mount(
      <MockedProvider>
        <Pagination
          fetchPage={fetchPage}
          itemsOnCurrentPage={10}
          itemsPerPage={10}
          itemsTotal={100}
          startIndex={10}
        />
      </MockedProvider>
    );

    const previousButton = wrapper
      .find('button[data-test="pagination-next"]')
      .first();
    previousButton.simulate("click");
    expect(fetchPage).toHaveBeenCalledWith(20, 10, "NEXT");
  });
});
