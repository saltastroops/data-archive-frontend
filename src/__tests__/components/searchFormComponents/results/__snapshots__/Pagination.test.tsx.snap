// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correctly with correct props 1`] = `
<Pagination
  fetchPage={[MockFunction]}
  itemsOnCurrentPage={100}
  itemsPerPage={100}
  itemsTotal={201}
  startIndex={0}
>
  <div
    className="pagination-controls"
  >
    <button
      className="pagination-previous"
      data-test="pagination-previous"
      disabled={true}
      onClick={[Function]}
    >
      Previous page
    </button>
    <a>
      <span
        className="pagination-ellipsis"
      >
        1
      </span>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
      <span
        className="pagination-ellipsis"
      >
        100
      </span>
      <span
        className="pagination-ellipsis"
      >
        of
      </span>
      <span
        className="pagination-ellipsis"
      >
        201
      </span>
    </a>
    <button
      className="pagination-next"
      data-test="pagination-next"
      disabled={false}
      onClick={[Function]}
    >
      Next page
    </button>
  </div>
</Pagination>
`;

exports[`Pagination should render correctly with correct props 2`] = `
<Pagination
  fetchPage={[MockFunction]}
  itemsOnCurrentPage={10}
  itemsPerPage={10}
  itemsTotal={201}
  startIndex={10}
>
  <div
    className="pagination-controls"
  >
    <button
      className="pagination-previous"
      data-test="pagination-previous"
      disabled={false}
      onClick={[Function]}
    >
      Previous page
    </button>
    <a>
      <span
        className="pagination-ellipsis"
      >
        11
      </span>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
      <span
        className="pagination-ellipsis"
      >
        20
      </span>
      <span
        className="pagination-ellipsis"
      >
        of
      </span>
      <span
        className="pagination-ellipsis"
      >
        201
      </span>
    </a>
    <button
      className="pagination-next"
      data-test="pagination-next"
      disabled={false}
      onClick={[Function]}
    >
      Next page
    </button>
  </div>
</Pagination>
`;

exports[`Pagination should render correctly with correct props 3`] = `
<Pagination
  fetchPage={[MockFunction]}
  itemsOnCurrentPage={201}
  itemsPerPage={1000}
  itemsTotal={201}
  startIndex={0}
>
  <div
    className="pagination-controls"
  >
    <button
      className="pagination-previous"
      data-test="pagination-previous"
      disabled={true}
      onClick={[Function]}
    >
      Previous page
    </button>
    <a>
      <span
        className="pagination-ellipsis"
      >
        1
      </span>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
      <span
        className="pagination-ellipsis"
      >
        201
      </span>
      <span
        className="pagination-ellipsis"
      >
        of
      </span>
      <span
        className="pagination-ellipsis"
      >
        201
      </span>
    </a>
    <button
      className="pagination-next"
      data-test="pagination-next"
      disabled={true}
      onClick={[Function]}
    >
      Next page
    </button>
  </div>
</Pagination>
`;
